File: android/domain/src/main/kotlin/com/bolsillo/domain/usecase/MoneyParser.kt
package com.bolsillo.domain.usecase

import com.bolsillo.domain.model.AmountInput
import com.bolsillo.domain.model.Currency
import com.bolsillo.domain.model.Money
import javax.inject.Inject

/**
 * Pure-integer keypad → [Money]. NO float/double path (Article III). The user
 * enters minor units directly (typing "100" on a 2-decimal currency means 1.00),
 * matching the canonical mock's amount strip.
 */
class MoneyParser
    @Inject
    constructor() {
    fun parse(
        input: AmountInput,
        currency: Currency,
    ): Money = parse(input.digits, currency)

    fun parse(
        digits: String,
        currency: Currency,
    ): Money {
        val clean = digits.trimStart('0')
        if (clean.isEmpty()) return Money.ZERO
        require(clean.all { it.isDigit() }) { "MoneyParser requires digit-only input" }
        require(currency.decimalDigits in 0..6) {
            "Unsupported decimalDigits=${currency.decimalDigits}"
        }
        return Money(clean.toLong())
    }

    /** Render a [Money] as the digit string the keypad would produce. */
    fun formatDigits(money: Money): String = if (money.minorUnits <= 0L) "" else money.minorUnits.toString()
}

File: ios/Packages/Bolsillo/Sources/BolsilloDomain/UseCases/MoneyParser.swift
import Foundation

/// Converts a raw digit string from the keypad into a `Money` value.
///
/// Uses integer math only — there is no Double/Float path (Constitution Article III).
/// The caller is responsible for supplying only digit characters; non-digit characters
/// are stripped defensively.
///
/// ## Minor-unit semantics
/// The digit string represents an amount already expressed in minor units.
/// For a 2-decimal-digit currency: "1234" → `Money(minorUnits: 1234)` → 12.34 display.
/// The display layer formats the value using `Currency.decimalDigits`.
public struct MoneyParser: Sendable {
    public init() {}

    /// Parse a string of digit characters into `Money`.
    ///
    /// - Parameters:
    ///   - digits: Raw keypad digits (e.g. "1234"). Non-digit characters are stripped.
    ///   - currency: The currency that determines decimal placement at display time.
    /// - Returns: `Money(minorUnits:)` for the parsed integer, or `.zero` on empty / invalid input.
    public func parse(digits: String, currency: Currency) -> Money {
        let raw = digits.filter(\.isNumber)
        guard !raw.isEmpty, let value = Int(raw) else { return .zero }
        return Money(minorUnits: value)
    }
}
